feature-ledger-impl/src/main/java/io/novafoundation/nova/feature_ledger_impl/sdk/application/substrate/newApp/NewSubstrateLedgerApplication.kt

```diff
@@ -0,0 +1,113 @@
+
+abstract class NewSubstrateLedgerApplication(
+    private val transport: LedgerTransport,
+    private val chainRegistry: ChainRegistry,
+    private val metadataShortenerService: MetadataShortenerService
+) : SubstrateLedgerApplication {
+
+    abstract val cla: UByte
+
+    abstract suspend fun getDerivationPath(chainId: ChainId, accountIndex: Int): String
+
+    abstract suspend fun getDerivationPath(metaId: Long, chainId: ChainId): String
+
+    override suspend fun getAccount(device: LedgerDevice, chainId: ChainId, accountIndex: Int, confirmAddress: Boolean): LedgerSubstrateAccount {
+        val displayVerificationDialog = SubstrateLedgerAppCommon.DisplayVerificationDialog.fromBoolean(confirmAddress)
+
+        val derivationPath = getDerivationPath(chainId, accountIndex)
+        val encodedDerivationPath = encodeDerivationPath(derivationPath)
+
+        val chain = chainRegistry.getChain(chainId)
+        val ss58Prefix = chain.addressPrefix.toShort()
+
+        val payload = encodedDerivationPath + ss58Prefix.littleEndianBytes
+
+        val rawResponse = transport.send(
+            cla = cla,
+            ins = SubstrateLedgerAppCommon.Instruction.GET_ADDRESS.code,
+            p1 = displayVerificationDialog.code,
+            p2 = defaultCryptoScheme().code,
+            data = payload,
+            device = device
+        )
+
+        Log.w("Ledger", "Got response (${rawResponse.size} bytes): ${rawResponse.joinToString()}")
+
+        return parseAccountResponse(rawResponse, derivationPath)
+    }
+
+    override suspend fun getSignature(
+        device: LedgerDevice,
+        metaId: Long,
+        chainId: ChainId,
+        payload: SignerPayloadExtrinsic
+    ): SignatureWrapper {
+        val chunks = prepareExtrinsicChunks(metaId, chainId, payload)
+
+        val results = chunks.mapIndexed { index, chunk ->
+            val chunkType = SubstrateLedgerAppCommon.SignPayloadType(index, chunks.size)
+
+            val rawResponse = transport.send(
+                cla = cla,
+                ins = SubstrateLedgerAppCommon.Instruction.SIGN.code,
+                p1 = chunkType.code,
+                p2 = defaultCryptoScheme().code,
+                data = chunk,
+                device = device
+            )
+
+            SubstrateLedgerAppCommon.processResponseCode(rawResponse)
+        }
+
+        val signatureWithType = results.last()
+
+        return SubstrateLedgerAppCommon.parseSignature(signatureWithType)
+    }
+
+    private suspend fun prepareExtrinsicChunks(
+        metaId: Long,
+        chainId: ChainId,
+        payload: SignerPayloadExtrinsic
+    ): List<ByteArray> {
+        val payloadBytes = payload.encodedSignaturePayload(hashBigPayloads = false)
+
+        val derivationPath = getDerivationPath(metaId, chainId)
+        val encodedDerivationPath = encodeDerivationPath(derivationPath)
+
+        val encodedTxPayloadLength = payloadBytes.size.toShort().littleEndianBytes
```

Comment:
Compared to legacy app chunk encoding, we also prepend tx blob length and append metadata proof :
https://github.com/Zondax/ledger-substrate-js/blob/main/src/common.ts#L223